src/engine/HasRelationScan.cpp

```diff
@@ -0,0 +1,458 @@
+  result->_resultTypes.push_back(ResultTable::ResultType::KB);
+  result->_resultTypes.push_back(ResultTable::ResultType::KB);
+  std::vector<std::array<Id, 2>>* fixedSizeData =
+      new std::vector<std::array<Id, 2>>();
+  result->_fixedSizeData = fixedSizeData;
+  fixedSizeData->reserve(resultSize);
+
+  size_t id = 0;
+  while (id < hasPattern.size() || id < hasRelation.size()) {
+    if (id < hasPattern.size() && hasPattern[id] != NO_PATTERN) {
+      // add the pattern
+      size_t numPredicates;
+      Id* patternData;
+      std::tie(patternData, numPredicates) = patterns[hasPattern[id]];
+      for (size_t i = 0; i < numPredicates; i++) {
+        fixedSizeData->push_back({{id, patternData[i]}});
+      }
+    } else if (id < hasRelation.size()) {
+      // add the relations
+      size_t numPredicates;
+      Id* predicateData;
+      std::tie(predicateData, numPredicates) = hasRelation[id];
+      for (size_t i = 0; i < numPredicates; i++) {
+        fixedSizeData->push_back({{id, predicateData[i]}});
+      }
+    }
+    id++;
+  }
+}
+
+template <typename T, typename C>
+struct resizeIfVec {
+  static void resize(T& t, int size) {
+    (void)t;
+    (void)size;
+  }
+};
+
+template <typename C>
+struct resizeIfVec<vector<C>, C> {
+  static void resize(vector<C>& t, int size) { t.resize(size); }
+};
+
+template <typename A, typename R>
+void doComputeSubqueryS(const std::vector<A>* input,
+                        const size_t inputSubjectColumn, std::vector<R>* result,
+                        const std::vector<PatternID>& hasPattern,
+                        const CompactStringVector<Id, Id>& hasRelation,
+                        const CompactStringVector<size_t, Id>& patterns) {
+  for (size_t i = 0; i < input->size(); i++) {
+    const A& inputRow = (*input)[i];
+    size_t id = inputRow[inputSubjectColumn];
+    if (id < hasPattern.size() && hasPattern[id] != NO_PATTERN) {
+      // add the pattern
+      size_t numPredicates;
+      Id* patternData;
+      std::tie(patternData, numPredicates) = patterns[hasPattern[id]];
+      for (size_t i = 0; i < numPredicates; i++) {
+        R row;
+        resizeIfVec<R, Id>::resize(row, inputRow.size() + 1);
+        for (size_t j = 0; j < inputRow.size(); j++) {
+          row[j] = inputRow[j];
+        }
+        row[inputRow.size()] = patternData[i];
+        result->push_back(row);
+      }
+    } else if (id < hasRelation.size()) {
+      // add the relations
+      size_t numPredicates;
+      Id* predicateData;
+      std::tie(predicateData, numPredicates) = hasRelation[id];
+      for (size_t i = 0; i < numPredicates; i++) {
+        R row;
+        resizeIfVec<R, Id>::resize(row, inputRow.size() + 1);
+        for (size_t j = 0; j < inputRow.size(); j++) {
+          row[j] = inputRow[j];
+        }
+        row[inputRow.size()] = predicateData[i];
+        result->push_back(row);
+      }
+    } else {
+      break;
+    }
+  }
+}
+
+/**
```

Comment:
We could do something similar for some of the other operations that currently use such a structure of nested ifs, right?

Reply:
We could use template meta programming for all of them, yes.